lib/datadog/tracing/span_operation.rb

```diff
@@ -383,15 +378,24 @@ def initialize
 
         # Triggered when the span raises an error during measurement.
         class OnError < Tracing::Event
-          def initialize
+          def initialize(default)
             super(:on_error)
+
+            @default = default
+            subscribe(&default)
           end
 
           # Call custom error handler but fallback to default behavior on failure.
+
+          # DEV: Revisit this before full 1.0 release.
+          # It seems like OnError wants to behave like a middleware stack,
+          # where each "subscriber"'s executed is chained to the previous one.
+          # This is different from how {Tracing::Event} works, and might be incompatible.
           def wrap_default(error_handler)
             return unless error_handler && error_handler.is_a?(Proc)
 
-            wrap(:default) do |original, op, error|
+            unsubscribe_all!
```

Comment:
Instead of unsubscribing, can we just do...?

```ruby
original = @default

@default = proc do |op, error|
  begin
    error_handler.call(op, error)
  rescue StandardError => e
    Datadog.logger.debug do
      "Custom on_error handler failed, using fallback behavior. \
      Error: #{e.message} Location: #{e.backtrace.first}"
    end

    original.call(op, error) if original
  end
end
```

Reply:
I change it to pretty much your suggestion.
I ended up dropping the `Tracing::Event` parent class because it's not possible to sanely support `subscribe` and `wrap_default` in a sane way. I got super confused when writing tests that supported both.
In my proposed changes, it now only supports the wrap method, making it look like a middleware stack.